src/lib/scheduler/abstract_task.cpp

```diff
@@ -108,6 +145,20 @@ void AbstractTask::_mark_as_scheduled() {
   DebugAssert((!already_scheduled), "Task was already scheduled!");
 }
 
+void AbstractTask::_on_exception_in_predecessor(const std::exception_ptr exception) {
+  _exception = exception;
+  _mark_as_done();
+  for (auto& successor : _successors) {
+    successor->_on_exception_in_predecessor(exception);
+  }
+
+  if (CurrentScheduler::is_set()) {
+    // update number of tasks executed so that Scheduler::finish finds the correct number of executed threads
+    auto worker = Worker::get_this_thread_worker();
+    worker->processing_unit().lock()->on_worker_finished_task();
```

Comment:
For diamond shapes, this will lead to `on_worker_finished_task()` getting called multiple times for a single task.